twilight-gateway/src/ratelimiter.rs

```diff
@@ -1,127 +1,240 @@
-    let remainder = RESET_DURATION_MILLISECONDS % heartbeat_interval;
+/// Ratelimit permit.
+///
+/// Holding one of these means there's capacity for sending *one* event.
+pub(crate) struct Permit<'a> {
+    /// `true` if the permit was unused.
+    unused: bool,
+    /// Reference to the ratelimiter.
+    inner: &'a mut CommandRatelimiter,
+}
 
-    // If we have a remainder then we reserve an additional heartbeat.
-    //
-    // If there is a remainder per reset then in theory we could allot one less
-    // command for heartbeating variably every number of resets, but it's best
-    // to be cautious and keep it simple.
-    if remainder > 0 {
-        heartbeats = heartbeats.saturating_add(1);
+impl<'a> Permit<'a> {
+    /// Create a new ratelimit permit.
+    fn new(ratelimiter: &'a mut CommandRatelimiter) -> Self {
+        Self {
+            unused: false,
+            inner: ratelimiter,
+        }
     }
 
-    // Convert the heartbeats to a u8. The number of heartbeats **should** never
-    // be above `u8::MAX`, so the error pattern branch should never be reached.
-    let heartbeats_converted = if let Ok(value) = heartbeats.try_into() {
-        value
-    } else {
-        tracing::warn!(
-            default=ALLOT_ON_FAIL,
-            %heartbeats,
-            "heartbeats > u8 max; defaulting",
-        );
+    /// Forget the permit, returning it to the [`CommandRatelimiter`].
+    ///
+    /// Use this when no event was sent.
+    pub(crate) fn forget(mut self) {
+        self.unused = true;
+    }
+}
 
-        ALLOT_ON_FAIL
-    };
+impl Drop for Permit<'_> {
+    fn drop(&mut self) {
+        if !self.unused {
+            self.inner.instants.push(Instant::now());
+        }
+    }
+}
 
-    COMMANDS_PER_RESET.saturating_sub(heartbeats_converted * 2)
+/// Calculate the number of non reserved commands for heartbeating (which skips
+/// the ratelimiter) in a given [`RESET_DURATION`].
+///
+/// Reserves capacity for the amount of heartbeats + 1, to account for Discord
+/// absurdly sending [`OpCode::Heartbeat`]s when the gateway is ratelimited
+/// (which requires the gateway to immediately send a heartbeat back).
+///
+/// [`OpCode::Heartbeat`]: twilight_model::gateway::OpCode::Heartbeat
+fn nonreserved_commands_per_reset(heartbeat_interval: Duration) -> u8 {
+    /// Guard against faulty gateway implementations sending absurdly low
+    /// heartbeat intervals by maximally reserving some number of heartbeats per
+    /// [`RESET_DURATION`].
+    const MAX_NONRESERVED_COMMANDS_PER_RESET: u8 = COMMANDS_PER_RESET - 10;
+
+    // Round up to be on the safe side.
+    #[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
+    let heartbeats_per_reset =
+        (RESET_DURATION.as_secs_f32() / heartbeat_interval.as_secs_f32()).ceil() as u8;
+
+    COMMANDS_PER_RESET
+        .saturating_sub(heartbeats_per_reset + 1)
+        .max(MAX_NONRESERVED_COMMANDS_PER_RESET)
```

Comment:
```suggestion
    // Calculate the amount of heartbeats per reset duration.
    let heartbeats_per_reset = RESET_DURATION.as_secs_f32() / heartbeat_interval.as_secs_f32();

    // Round up to be on the safe side.
    #[allow(clippy::cast_sign_loss, clippy::cast_possible_truncation)]
    let heartbeats_per_reset = heartbeats_per_reset.ceil() as u8;

    // Reserve an extra heartbeat just in case.
    let heartbeats_per_reset = heartbeats_per_reset + 1;

    // Subtract the reserved heartbeats from the total available commands.
    let nonreserved_commands_per_reset = COMMANDS_PER_RESET.saturating_sub(heartbeats_per_reset);

    // Take the larger value between this and our guard value.
    nonreserved_commands_per_reset.max(MAX_NONRESERVED_COMMANDS_PER_RESET)
```